File: php/examples/message.php
<?php
/*
   OpenChange PHP bindings examples

   Copyright (C) 2013-2014 Javier Amor Garcia

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

include('./test-helpers.php');
include('./config.php');

$pidTagBody1 = 'body1';
$pidTagDisplayName1 = 'displayName1';
$pidLidWorkAddressStreet1 = 'street1';
$pidLidEmail1EmailAddress1 = 'jk1@example1.org';
$pidTagBody2 = 'body2';
$pidTagDisplayName2 = 'displayName2';
$pidLidWorkAddressStreet2 = 'street2';
$pidLidEmail1EmailAddress2 = 'jk2@example2.org';

$mapi = new MAPIProfileDB($dbPath);
ok($mapi, "MAPIProfileDB opened from $dbPath");

$mapiProfile = $mapi->getProfile($profileName);
ok($mapiProfile, "Profile $profileName");

$session = $mapiProfile->logon();
ok($session, "Logon with $profileName");

$mailbox = $session->mailbox();
ok($mailbox, "Default mailbox for $profileName");

$contacts = $mailbox->contacts();
ok($contacts, "Opening contact folder");

ok(is_null($contacts->openMessage($inexistentMesageId)), "Trying to open a inexistent message return NULL");

$message = $contacts->openMessage($contactMessageId,  MAPIMessage::RW);
ok($message, "Open message RW");
is($message->getID(), $contactMessageId, "Checking opening message ID");
is($message->getBodyContentFormat(), "txt", "Checking body content format");

diag("Setting multiple messages properties to first set");
$message->set(
	PidTagBody, $pidTagBody1,
	PidLidWorkAddressStreet, $pidLidWorkAddressStreet1,
	PidLidEmail1EmailAddress, $pidLidEmail1EmailAddress1
);

diag("Setting one property PidTagDisplayName to first set");
$message->set(PidTagDisplayName, $pidTagDisplayName1);

$message->save();
unset($message);

diag ("- Saving changes");
$message = $contacts->openMessage($contactMessageId, MAPIMessage::RW);
ok($message, "Reopening message in RW mode");

is($message->get(PidTagDisplayName), $pidTagDisplayName1, "Checking PidTagDisplayName1 has changed properly");
$multipleValues = $message->get(PidTagBody, PidLidWorkAddressStreet, PidLidEmail1EmailAddress);
is($multipleValues[PidTagBody], $pidTagBody1, "Checking value as PidTagBody (retrieved in a multiple value get)");
is($multipleValues[PidLidWorkAddressStreet], $pidLidWorkAddressStreet1, "Checking value as PidLidWorkAddressStreet (retrieved in a multiple value get)");
is($multipleValues[PidLidEmail1EmailAddress], $pidLidEmail1EmailAddress1, "Checking value as PidLidEmail1EmailAddress (retrieved in a multiple value get)");

diag("Setting values again");
$message->set(
	PidTagBody , $pidTagBody2,
	PidLidWorkAddressStreet , $pidLidWorkAddressStreet2,
	PidLidEmail1EmailAddress , $pidLidEmail1EmailAddress2
);
$message->set(PidTagDisplayName, $pidTagDisplayName2);

$message->save();
unset($message);

diag("- Saving last changes");

$message = $contacts->openMessage($contactMessageId, MAPIMessage::RO);
ok($message, "Reopening message in RO mode");

is($message->get(PidTagDisplayName), $pidTagDisplayName2, "Checking PidTagDisplayName2 has changed properly");
$multipleValues = $message->get(PidTagBody, PidLidWorkAddressStreet, PidLidEmail1EmailAddress);
is($multipleValues[PidTagBody], $pidTagBody2, "Checking value as PidTagBody (retrieved in a multiple value get)");
is($multipleValues[PidLidWorkAddressStreet], $pidLidWorkAddressStreet2, "Checking value as PidLidWorkAddressStreet (retrieved in a multiple value get)");
is($multipleValues[PidLidEmail1EmailAddress], $pidLidEmail1EmailAddress2, "Checking value as PidLidEmail2EmailAddress (retrieved in a multiple value get)");


unset($message);

endTestSuite("message.php");

?>

File: libmapi++/src/message.cpp
/*
   libmapi C++ Wrapper
   Message class implementation

   Copyright (C) Alan Alvarez 2008.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <libmapi++/attachment.h>

namespace libmapipp {

message::attachment_container_type message::fetch_attachments()
{
	mapi_object_t   attachment_table;

	mapi_object_init(&attachment_table);
	if (GetAttachmentTable(&m_object, &attachment_table) != MAPI_E_SUCCESS) {
		mapi_object_release(&attachment_table);
		throw mapi_exception(GetLastError(), "message::fetch_attachments : GetAttachmentTable");
	}

	SPropTagArray* property_tag_array = set_SPropTagArray(m_session.get_memory_ctx(), 0x1, PR_ATTACH_NUM);

	if (SetColumns(&attachment_table, property_tag_array) != MAPI_E_SUCCESS) {
		MAPIFreeBuffer(property_tag_array);
		mapi_object_release(&attachment_table);
		throw mapi_exception(GetLastError(), "message::fetch_attachments : SetColumns");
	}

	MAPIFreeBuffer(property_tag_array);

	SRowSet  row_set;
	attachment_container_type attachment_container;

	while( (QueryRows(&attachment_table, 0x32, TBL_ADVANCE, TBL_FORWARD_READ, &row_set) == MAPI_E_SUCCESS) && row_set.cRows) {
		for (unsigned int i = 0; i < row_set.cRows; ++i) {
			try {
				attachment_container.push_back(attachment_shared_ptr(new attachment(*this, row_set.aRow[i].lpProps[0].value.l)));
			}
			catch(mapi_exception e) {
				mapi_object_release(&attachment_table);
				throw;
			}
		}
	}
	mapi_object_release(&attachment_table);

	return attachment_container;
}

} // namespace libmapipp
